src/main/java/org/commcare/core/network/AuthenticationInterceptor.java

```diff
@@ -0,0 +1,43 @@
+package org.commcare.core.network;
+
+import java.io.IOException;
+
+import javax.annotation.Nullable;
+
+import okhttp3.Interceptor;
+import okhttp3.Request;
+import okhttp3.Response;
+
+public class AuthenticationInterceptor implements Interceptor {
+
+    @Nullable
+    private String credential;
+
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+        Request original = chain.request();
+
+        // if credentials are null, proceed without authenticating
+        if (credential == null) {
+            return chain.proceed(original);
+        }
+
+        // Throw an exception if we are sending an authenticated request over HTTP
+        if (!original.isHttps() && credential != null) {
+            throw new PlainTextPasswordException();
+        }
+
+        Request.Builder builder = original.newBuilder()
+                .header("Authorization", credential);
+
+        Request request = builder.build();
+        return chain.proceed(request);
+    }
+
+    public void setCredential(@Nullable String credential) {
+        this.credential = credential;
+    }
+
+    public static class PlainTextPasswordException extends RuntimeException {
```

Comment:
Seems a bit odd for this to be a RuntimeException. I'd imagine that this should be an IOException base?

Reply:
Hmm. May be it makes sense here because our request urls are dynamic and dependent on the ccz profile rather than the app structure. Though I just copied it from earlier implementation and is not sure of the original intent.